plasmapy/plasma/grids.py

```diff
@@ -0,0 +1,330 @@
+"""
+Defines the AbstractGrid class and child classes
+"""
+
+__all__ = [
+    "AbstractGrid",
+    "CartesianGrid",
+]
+
+from abc import ABC
+import astropy.units as u
+import numpy as np
+
+
+class AbstractGrid(ABC):
+    """
+    Abstract grid represents a 3D grid of positions. The grid is stored as an
+    np.ndarray, while the units associated with each dimension are stored
+    separately.
+    """
+
+    # TODO: add appropriate typing here on start, stop, num
+    # start/stop can be -> number(int,float), ndarray, u.quantity, list of numbers
+    # or u.quantities units can be astropy.units.core.Unit or list of same
+    def __init__(self, start=0, stop=1, num=100, grid=None, units=None):
```

Comment:
`start` and `stop` should be required arguments and, thus, no default values.

Reply:
As currently written they cannot be required because they should be optional in the case that the 'grid' keyword is set (loading a user-created grid rather than creating a new one). Should we have them all default to None then enforce this via a check, or is there a better way?

Reply:
How about something like...

```python
def __init__(self, *seeds, num=100, units=None, **kwargs):
    if len(seeds) == 1:
        # assume grid
        self._load_grid(seeds[0], units=units)
    elif len(seeds) == 2:
        # assume start-stop
        self._make_grid(seeds[0], sees[1], num=num, **kwargs)
    else:
        raise TypeError(f"{self.__class__.__name}() takes 1 or 2 positional arguments but {len(seeds)} were given")

def _load_grid(self, grid, units=None):
    if not isinstance(grid, self.__class__):
        raise VelueError(f"Expected instance of {self.__class__} and got {grid.__class__}."

def _make_grid(self, start, stop, num=100, **kwargs):

```
Notes:  I did drop the `grid` kwarg on purpose and the `grid` input to `_load_grid` should be a required argument.

Reply:
Sounds good: done

Reply:
Then we can get rid of the TODO note, right?